### PR TITLE
[ENH] Added tonic input to GUI drivers tab

### PR DESCRIPTION
**Pull request by kmilo9999**
Wok on new Feature to add tonic input from in the GUI
Solves #715 

image
image

### CODE REVIEW

**Comment by gtdang** [comment]
@ntolley @jasmainak  @rythorpe  Thoughts on if times should be above or below the amplitudes?

**Comment by rythorpe** [comment]
Correct, synchronous inputs don't make sense to tonic inputs.

File: hnn_core/gui/gui.py [hunk 1]
```diff
@@ -1319,6 +1390,17 @@ def _init_network_from_widgets(params, dt, tstop, single_simulation_data,
...
+        if drive['type'] in ('Tonic'):
+            weights_amplitudes = {
+                k: v.value
+                for k, v in drive["amplitude"].items()
+            }
+            single_simulation_data['net'].add_tonic_bias(
+                amplitude=weights_amplitudes,
+                t0=drive["t0"].value,
+                tstop=drive["tstop"].value)
+            return
```
Comment (gtdang) [inline hnn_core/gui/gui.py:1402]: Should this be returning?

**Review by gtdang — CHANGES_REQUESTED** [review]
Looking good Camilo! Per the discussion:

- [x] Remove Sync Inputs checkbox from tonics 
- [ ] Move times above amplitudes 
- [ ] Add test to check tonic widget is created with prespecified_drive_data 
- [ ] Add test to check an added tonic amplitude is added to the network when _init_network_from_widgets is called

File: hnn_core/gui/gui.py (hunk 1 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py:1402]: The rest of the logic in that function doesnt apply to Tonic inputs. I want to avoid the `elif` branching

Comment (kmilo9999) [reply hnn_core/gui/gui.py:1402]: We talked it over and I'll remove the return

File: hnn_core/gui/gui.py [hunk 2]
```diff
@@ -1512,5 +1588,9 @@ def launch():
     You can pass voila commandline parameters as usual.
     """
     from voila.app import main
+    import debugpy
+    debugpy.listen(5678)
+    print("Waiting for debugger attach")
+    debugpy.wait_for_client()
```
Comment (gtdang) [inline hnn_core/gui/gui.py:1594]: Debug lines snuck into PR.

**Review by gtdang — COMMENTED** [review]
I played around with this and replicated the [Simulate Gamma API Tutorial](https://jonescompneurolab.github.io/hnn-core/stable/auto_examples/workflows/plot_simulate_gamma.html#sphx-glr-auto-examples-workflows-plot-simulate-gamma-py) tonic example. Below is the plot comparing the simulations with and without the added tonic. 
Screenshot 2024-05-22 at 11 09 01 AM
I think everything is working as expected! Great job Camilo! 

Some quality of life considerations. 
* When adding a tonic drive, the default _Stop Time_ is set to 0. Should this sync with the simulation tstop parameter on creation?
* The simulation tab stop time label is "tstop (ms)" while in the tonic it's "Stop Time". Should we be consistent here? And which do you prefer? 
* I think the cell types should be ordered alphabetically: [L2_basket, L2_pyramidal, L5_basket, L5_pyramidal]

@ntolley @jasmainak @rythorpe  Thoughts?

**Comment by kmilo9999** [comment]
image

File: hnn_core/gui/gui.py [hunk 3]
```diff
@@ -457,6 +458,10 @@ def _simulation_list_change(value):
                     is_disabled="", btn_height=self.layout['run_btn'].height,
                     color_theme=self.layout['theme_color']))
 
+        def _driver_type_change(value):
+            self.widget_location_selection.disabled = (
```
Comment (ntolley) [inline hnn_core/gui/gui.py:462]: Very useful! Glad to see this is implemented natively

File: hnn_core/gui/gui.py [hunk 4]
```diff
@@ -1015,6 +1019,58 @@ def _get_evoked_widget(name, layout, style, location, data=None,
     return drive, drive_box
 
 
+def _get_tonic_widget(name, layout, style, data=None):
+
+    default_data = {
+        'amplitude': 0,
+        't0': 0,
+        'tstop': 0,
+        'seedcore': 14,
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1028]:
```suggestion
```
is `seedcore` actually used anywhere? `net.add_tonic_drives` doesn't have any randomization so probably safe to delete

File: hnn_core/gui/gui.py [hunk 5]
```diff
@@ -1015,6 +1019,58 @@ def _get_evoked_widget(name, layout, style, location, data=None,
...
+    for cell_type in cell_types:
+        if cell_type in data:
+            default_data.update(data[cell_type])
+        amplitudes[cell_type] = BoundedFloatText(
+            value=deepcopy(default_data['amplitude']),
+            description=cell_type, min=0, max=1e6, step=0.01, **kwargs)
+
+    start_times = BoundedFloatText(
+        value=deepcopy(default_data['t0']), description="Start time",
+        min=0, max=1e6, step=0.01, **kwargs)
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1047]:
```suggestion
        min=0, max=1e6, step=1.0, **kwargs)
```
This is a pretty small step size considering what's being changed

File: hnn_core/gui/gui.py [hunk 6]
```diff
@@ -1015,6 +1019,58 @@ def _get_evoked_widget(name, layout, style, location, data=None,
...
+    stop_times = BoundedFloatText(
+        value=deepcopy(default_data['tstop']), description="Stop time",
+        min=-1, max=1e6, step=0.01, **kwargs)
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1050]:
```suggestion
        min=-1, max=1e6, step=1.0, **kwargs)
```

File: hnn_core/gui/gui.py [hunk 7]
```diff
@@ -1026,6 +1082,10 @@ def add_drive_widget(drive_type, drive_boxes, drive_widgets, drives_out,
                      sync_evinput=False):
     """Add a widget for a new drive."""
 
+    # Check only adds 1 tonic input widget
+    if drive_type == "Tonic" and not _is_valid_add_tonic_input(drive_widgets):
+        return
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1087]: Does this just mean the add drive button doesn't do anything once you add a tonic drive?

**Comment by gtdang** [comment]
> I played around with this and replicated the [Simulate Gamma API Tutorial](https://jonescompneurolab.github.io/hnn-core/stable/auto_examples/workflows/plot_simulate_gamma.html#sphx-glr-auto-examples-workflows-plot-simulate-gamma-py) tonic example. Below is the plot comparing the simulations with and without the added tonic. Screenshot 2024-05-22 at 11 09 01 AM I think everything is working as expected! Great job Camilo!
> 
> Some quality of life considerations.
> 
> * When adding a tonic drive, the default _Stop Time_ is set to 0. Should this sync with the simulation tstop parameter on creation?
> * The simulation tab stop time label is "tstop (ms)" while in the tonic it's "Stop Time". Should we be consistent here? And which do you prefer?
> * I think the cell types should be ordered alphabetically: [L2_basket, L2_pyramidal, L5_basket, L5_pyramidal]
> 
> @ntolley @jasmainak @rythorpe Thoughts?

This was discussed at our dev sync. 
1.) We are not syncing tstop. Poisson drive does not. 
2.) Will open an issue to standardize name and unit labels 
3.) Cell types are ordered alphabetically now.

File: hnn_core/gui/gui.py (hunk 7 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py:1087]: That is correct.
@gtdang

File: hnn_core/gui/gui.py (hunk 6 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Done

File: hnn_core/gui/gui.py (hunk 5 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Done

File: hnn_core/gui/gui.py (hunk 4 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Removed

**Comment by kmilo9999** [comment]
@gtdang @ntolley The CodeSpell checks fails in a file I have not modified in this PR

**Review by ntolley — APPROVED** [review]
Great work @kmilo9999 !!

I'm a bit confused about the codespell, I wonder if the github action somehow got updated and is now checking new files...

Do you think you could modify the typos in any case? It's just the 3 lines and they'll need to be fixed at some point

**Comment by gtdang** [comment]
> Great work @kmilo9999 !!
> 
> I'm a bit confused about the codespell, I wonder if the github action somehow got updated and is now checking new files...
> 
> Do you think you could modify the typos in any case? It's just the 3 lines and they'll need to be fixed at some point

Created a separate PR for those typos #777

**Comment by ntolley** [comment]
@gtdang is this request satisfied?

> - [ ]  Add test to check tonic widget is created with prespecified_drive_data

Did you mean for there to be a test with `assert` statements?

**Comment by kmilo9999** [comment]
> @gtdang is this request satisfied?
> 
> > * [ ]   Add test to check tonic widget is created with prespecified_drive_data
> 
> Did you mean for there to be a test with `assert` statements?

This `prespecified_drive_data` function reads the params from `default.json`, and I dont see a tonic bias input described on this file. Also, I think some logic needs to be implemented to be able to read and load tonic bias from a file.

**Comment by gtdang** [comment]
> > @gtdang is this request satisfied?
> > > * [ ]    Add test to check tonic widget is created with prespecified_drive_data
> > 
> > 
> > Did you mean for there to be a test with `assert` statements?
> 
> This `prespecified_drive_data` function reads the params from `default.json`, and I dont see a tonic bias input described on this file. Also, I think some logic needs to be implemented to be able to read and load tonic bias from a file.

@ntolley is there an example of the flat a json or param file with tonics encoded?

**Comment by kmilo9999** [comment]
> These are the parameters in the `.param` file that need to be modified image
> 
> Do you think we should save loading for a separate PR and go ahead with merging this one?

@ntolley why the L2Basket and L5Basket cell types dont have the suffix `_soma`?

**Comment by ntolley** [comment]
> @ntolley why the L2Basket and L5Basket cell types dont have the suffix _soma?

These cells are modeled as just one single `Section`, so the whole cell is represented by the soma.

**Comment by ntolley** [comment]
Merging since every test is passing besides Ubuntu (which is in progress)

Thanks @kmilo9999!!!

**Pull request merged**